### ISSUE TITLE
How to modify original entity during audit save

### ISSUE THREAD

**Issue by nologinatgit** [thepirat000/Audit.NET#211]
I have the following setup to save audit information about the opt-in tabled:

```csharp
Audit.Core.Configuration.Setup()
.UseEntityFramework(_ => _.AuditTypeMapper(t => typeof(DBInfo))
.AuditEntityAction<DBInfo>(async (ev, entity, auditEntity) =>
{
 auditEntity.ID = Guid.NewGuid();
 auditEntity.Time = DateTime.UtcNow;           
 auditEntity.Table = entity.Table;
 auditEntity.EventType = entity.Action;
 auditEntity.User = await dataProvider.GetCurrentUser(); //always null after saving...strange
}).IgnoreMatchedProperties(true));
```

I intend to modify those tables during audit save to avoid having to modify each and every one of them during regular modifications.

My idea was the following: when the audit entity is about to be saved, I modify the original entity and on `SaveChanges` it gets persisted.

However, when I added the following code, I get an _**infinite loop**_: the `AuditEntityAction` keeps being called.

```csharp
Audit.Core.Configuration.AddCustomAction(ActionType.OnEventSaving, async scope =>
{
    //Audit.Core.Configuration.AuditDisabled = true; //it brakes the saving...
    var efEvent = scope.GetEntityFrameworkEvent();
    if (efEvent != null)
    {
      // this is an EF event, set the eventtype before saving assuming only one Entry          
      dynamic originalEntity = efEvent.Entries.First().Entity;
      originalEntity.DataCount += 1; 
      originalEntity.CurrentDate = DateTime.Now;
      originalEntity.User = await dataProvider.GetCurrentUser(); //always null after saving...strange
    }
    //Audit.Core.Configuration.AuditDisabled = false; //it brakes the saving...
});
```

What else would you suggest to use if I don't want to modify every audited entity?

**Comment by thepirat000** [comment]
I'm not sure what do you mean by:

> My idea was the following: when the audit entity is about to be saved, I modify the original entity and on SaveChanges it gets persisted.

About your second example, the `OnEventSaving` action occurs AFTER the EF entity audited is modified, so the changes you are doing on the "original entity" will not affect the operation you are auditing. 

If you want to update/calculate fields on the Audit Entity before saving it, the `AuditEntityAction` is the way to go.

You didn't show what the `dataProvider` variable is and how it is created/disposed, I cannot tell why it's returning null

**Comment by nologinatgit** [comment]
I give some explanation.

I have several entities (e.g. `Data1`, `Data2`, `Data3 ` and so on) and each of them has the `DataCount`, `CurrentDate` and `User` fields.

I want those fields to be filled _automatically_.

I have two options:

1.  I modify the `Save` method inside `Data1Provider` to set the fields, and I modify the `Save `method inside `Data2Provider`, etc.
2. I get an event before they are saved to the database and insert the values just right before the universal saving

The second option is where Audit.Net comes into picture. As it is hooked around the base `SaveChanges`, I tried to set the values there.

So what would you recommend?

Should this work?

```csharp
Audit.Core.Configuration.Setup()
.UseEntityFramework(_ => _.AuditTypeMapper(t => typeof(DBInfo))
.AuditEntityAction<DBInfo>(async (ev, entity, auditEntity) =>
{
 auditEntity.ID = Guid.NewGuid();
 auditEntity.Time = DateTime.UtcNow;           
 auditEntity.Table = entity.Table;
 auditEntity.EventType = entity.Action;
 auditEntity.User = await dataProvider.GetCurrentUser(); //always null after saving...strange
 if(entity.Entity != null)
 {
      dynamic originalEntity = entity.Entity;
      originalEntity.DataCount += 1; 
      originalEntity.CurrentDate = DateTime.Now;
      originalEntity.User = await dataProvider.GetCurrentUser(); 
 }
}).IgnoreMatchedProperties(true));
```

**Comment by thepirat000** [comment]
My recommendation is to use the library for its purpose, that is auditing changes on a DbContext, and not updating fields to the original entities on the auditing configuration.

Anyway, the only chance you have to manipulate the EF event **before** it's saved, is on `OnScopeCreated` action

```c#
Audit.Core.Configuration.AddCustomAction(ActionType.OnScopeCreated, scopee =>
{
    var efEvent = scopee.GetEntityFrameworkEvent();
    // ...
});
```

Or overriding the `OnScopeCreated` method on the `AuditDbContext`:

```c#
public class YourDbContext : AuditDbContext
{
    public override void OnScopeCreated(AuditScope auditScope)
    {
        var efEvent = auditScope.GetEntityFrameworkEvent();
        // ...
    }

}
```

**Comment by nologinatgit** [comment]
Two notes:

1. I thought that as Audit.NET encapsulates the `base.SaveChanges()` it is called before the original entity is persisted. That's why I wanted to modify the fields there.
2. Unfortunately, the `await dataprovider.dataProvider.GetCurrentUser();`  is an async method and it is not returned any result prior to the time when the audit entity is saved, so it is always saved as null.
I tried to add an `async` keyword to the Action deletegate but it is not awaited.
I tried even in `OnScopeCreated`, but is still null when the saving is done.

```csharp
Audit.Core.Configuration.AddCustomAction(ActionType.OnEventSaving, async scope =>
{
   userName = await dataProvider.GetCurrentUser;
}
```

**Comment by thepirat000** [comment]
Did you tried waiting synchronously like `userName = dataProvider.GetCurrentUser().GetAwaiter().GetResult();` ?

**Comment by thepirat000** [comment]
Also didn't understand why

> it is not returned any result prior to the time when the audit entity is saved, so it is always saved as null.

so you don't have the username until the audit is already saved?, that makes no sense

**Comment by nologinatgit** [comment]
I have the username inside a cache. As everything is modularized, I have to ask the cache for the username, and to avoid blocking calls, the cache is async.
I tried to call the `GetAwaiter().GetResult()` call, and it causes a dead lock (which is not rare among async calls, see [Don't block on async calls](https://blog.stephencleary.com/2012/07/dont-block-on-async-code.html).
As a workaround, I set a static value to contain the username...very poor solution though and not too safe as well.

I followed your suggestion and modified the original entity in the  `Audit.Core.Configuration.AddOnCreatedAction(async scope =>...)` part of my code. It seems to be working, I need to do some more tests.
